documentation/projects/proposals/rekognition_data/20240530-implementation_plan_augment_catalog_with_rekognition_tags.md

```diff
@@ -0,0 +1,669 @@
+- Existing Clarifai tags which do not match the above criteria will be removed
+  from the database
+- New Rekognition tags which do match the above criteria will be added to the
+  database
+
+## Label criteria
+
+This section describes the criteria used for determining which machine-generated
+tags we should exclude when adding any new tags to the database, and what the
+minimum accuracy cutoff for those tags should be.
+
+### Label selection
+
+<!-- Citations -->
+
+[^1]:
+    [N. Garcia, Y. Hirota, Y. Wu and Y. Nakashima, "Uncurated Image-Text Datasets: Shedding Light on Demographic Bias," _2023 IEEE/CVF Conference on Computer Vision and Pattern Recognition (CVPR)_, Vancouver, BC, Canada, 2023, pp. 6957-6966, doi: 10.1109/CVPR52729.2023.00672.](https://ieeexplore.ieee.org/document/10204859)
+
+[^2]:
+    [Schwemmer C, Knight C, Bello-Pardo ED, Oklobdzija S, Schoonvelde M, Lockhart JW. Diagnosing Gender Bias in Image Recognition Systems. Socius. 2020 Jan-Dec;6:10.1177/2378023120967171. doi: 10.1177/2378023120967171. Epub 2020 Nov 11. PMID: 35936509; PMCID: PMC9351609.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9351609/)
+
+[^3]:
+    [D. Zhao, A. Wang, and O. Russakovsky, "Understanding and evaluating racial biases in image captioning," _2021 IEEE/CVF International Conference on Computer Vision (ICCV)_, Oct. 2021, doi: 10.1109/iccv48922.2021.01456.](https://openaccess.thecvf.com/content/ICCV2021/papers/Zhao_Understanding_and_Evaluating_Racial_Biases_in_Image_Captioning_ICCV_2021_paper)
+
+[^4]:
+    [I. D. Raji and J. Buolamwini, “Actionable Auditing,” _MIT Media Lab_, Jan. 2019, doi: 10.1145/3306618.3314244.](https://www.aies-conference.com/2019/wp-content/uploads/2019/01/AIES-19_paper_223.pdf)
+
+[^5]:
+    [Bass, D. (2019, April 3). Amazon Schooled on AI Facial Technology By Turing Award Winner. _Bloomberg_.](https://www.bloomberg.com/news/articles/2019-04-03/amazon-schooled-on-ai-facial-technology-by-turing-award-winner)
+
+[^6]:
+    [Buolamwini, J. (2019, January 25). Response: Racial and Gender bias in Amazon Rekognition — Commercial AI System for Analyzing Faces. _Medium_.](https://medium.com/@Joy.Buolamwini/response-racial-and-gender-bias-in-amazon-rekognition-commercial-ai-system-for-analyzing-faces-a289222eeced)
+
+Machine-generated tags that are the product of AI image labeling models have
+been shown repeatedly and consistently to perpetuate certain cultural,
+structural, and institutional biases[^1][^2][^3]. This includes analysis done on
+[AWS Rekognition](https://docs.aws.amazon.com/rekognition/),
+specifically[^4][^5][^6].
+
+Certain demographic axes seem the most likely to result in an incorrect or
+insensitive label (e.g. gender assumption of an individual in a photo). For the
+reasons described in the above cited works, we should **exclude** labels that
+have a demographic context in the following categories:
+
+- Age
+- Gender
+- Sexual orientation
+- Nationality
+- Race
```

Comment:
I am completely on board with:

1. Treat the catalog as a data warehouse, store all/as much initial data as possible.
2. Do not surface _any_ demographic/identity/affiliation information about human subjects. 

I totally understand the intention behind "less likely to be applied in an insensitive manner" but I don't think it is a strong enough heuristic for use here.

Reply:
Thanks for both your input on this! As described in my comment above, I had just gotten my perspective locked into a certain view of the catalog and what was possible on the API. I appreciate your thoughts here, and I think I'd like to make the following changes to the approach:
1. Store all the data we can in the catalog
2. Filter the tags out as part of the indexing into Elasticsearch, leaving the `SELECT` from the catalog to the API as simple as possible
3. Only surface tags that are returned in the Elasticsearch hit, not all the tags that are available in the API database (since it will have them all)
4. Don't surface any demographic labels

This will give us flexibility for changing the data down the line (we only need to update the filters and reindex Elasticsearch, we don't need to rebuild the API DB from scratch), while preventing the need for adding steps to filter out each result in the API code and potentially affecting performance there.

Reply:
FWIW I think we should filter the tags when inserting in to the API database as well, otherwise every search response has an N*N+1 loop to filter the tags during serialization. We can amortise that by doing it when inserting into the API database, from which I believe the indexers select the Elasticsearch data to begin with, right? If we filter it when inserting into the API, then we don't need to _also_ filter it for Elasticsearch, and then don't need to do anything when returning a search.

Reply:
The filtering on insert into the API database is not something that we're currently set up to do currently, though. The place we do it during the data refresh is in the cleaning step, which we're currently removing as part of the data normalization process.

My suggestion was to to filter when indexing into Elasticsearch, then simply replace the `tags` attribute of each result returned from the API with the `tags` entry in the matching Elasticsearch document. Do you think that approach would be feasible? Otherwise we'd need to make this insert query much more complicated: https://github.com/WordPress/openverse/blob/b4ab20cdb220f442949d9c06a99ced5e80c1e1e1/ingestion_server/ingestion_server/queries.py#L168

Reply:
We could also consider switching to serving single results from Elasticsearch rather than the API DB to address this. I haven't considered the feasibility of it yet but we'd essentially just need to compare what data we don't index in es and if there would be downsides to adding it (slowing queries, filling the es disk, etc.)

Reply:
Adding my explicit +1 to what's been discussed, particularly removing demographic labels for human subjects. I had a note about religious affiliation as well.

Reply:
(I've referred to the catalogue DB as "Cat. DB" and Elasticsearch as ES in this).

> The filtering on insert into the API database is not something that we're currently set up to do currently, though. The place we do it during the data refresh is in the cleaning step, which we're currently removing as part of the data normalization process.

Yes, I recognise that. We (Staci, you and I) have just had a discussion about the different parts of the ETL pipeline, from provider to Elasticsearch, and I'm not entirely sure what the implications are for this now, but will share some further thoughts. Basically my suggestion overall is that perhaps attempting to entirely remove the T part of the ETL pipeline of Cat. DB -> API DB is unwise, because clearly we still need to sometimes do _some_ transformations after the catalog database. However, that is only true if the catalogue database really is the home for _everything_ at rest, including things we might not use right away, like tags we want to keep but don't want to use now.

One thing we discussed just now was potentially treating the Rekognition S3 -> Cat. DB ETL as not a one-time process, but something we could re-run in the future if we decided to change how we used the Rekogntion data. I'd like to formally suggest that now, because to my mind it would answer and simplify two rather urgent questions: first, on the purpose of the Cat. DB, API DB, and Elasticsearch as data destinations; second, on the purpose of the ETL between each of those. Pushing the filtering to before the labels reach the Cat. DB while definitively preserving the Rekognition data in S3, in part avoids these questions by putting the matter earlier in the data lifecycle, but also (most significantly) avoids introducing a new muddling of the purpose and relationship between the API DB and ES in particular.

On the matter of where to filter these specific values. Your suggestion to swap the ES tags into the API DB result when returning is clever, and gets rid of the `*N` issue of needing to filter individual tags during the API response. However, it sets a precedent that I'm uncomfortable with and which explicitly reverses decisions we've made in the past regarding the relationship between the API DB and ES. To put an even finer point on it, we used to have a muddied relationship between the two, where sometimes one or the other was used as the presentable data. Now, however, we have resolved that, and always use the API for presentation.  [See this PR as the notable inflection point of this relationship, which explicitly treats the API database as the presentable data, not ES](https://github.com/WordPress/openverse/pull/3408). It would, as Zack pointed out, also require a single result response to query both ES and the API database, whereas now it only queries the API database. In my opinion, we need to have a single place that represents the presentable data, and Elasticsearch isn't that, not since the PR I've mentioned, and we've in fact explicitly decided against that. In the current formulation of responsibility, Elasticsearch is for searching, and the API database is for presenting, and in fact includes additional data (to name some examples, waveform, license version, height and width of images used for oembed, audio set and genre for audio), which isn't in ES. Treating part of ES as the presentable data and part of the API database as the presentable data introduces mixed concerns between the two, which significantly complicates reasoning about how the API functions, and the purpose of each data location. This, to me, is critical to solve (and to prevent further complications to), and other than not including the data in the catalogue database (while making it possible to easily do so in the future, by re-running the Rekognition S3 -> Cat. DB ETL DAG), the only way to prevent that is to filter at the Cat. DB -> API DB stage, which you've deemed too complex (though I acknowledge Staci has a potentially differing opinion here), and which would be a new thing anyway. My suggestion avoids adding new things to existing pieces of our data puzzle until we have clearer overall expectations of the responsibilities of each part of our overall data lifecycle. This might also have the additional happy side effect of being easily reusable or adaptable to incorporating any data generated as part of #1968 (which is outside the scope of both these projects, but which we should anticipate).

Therefore, I suggest treating Rekognition S3 -> Cat. DB ETL as a repeatable process, and that the Rekognition data would stay at rest in S3 forever, and represent _all_ the Rekognition data (at least from this source of that data).

My suggestion is also rooted in an understanding that perhaps the Cat. DB could one day be the place we keep _everything_ regardless of its downstream purpose. But for now, because these questions are introducing new complications into an IP that was otherwise meant to be relatively simple, and because there's a way of "punting" the question down the road, I suggest doing so. Thereby, we can avoid setting somewhat new precedents or reversing important past decisions now which aren't necessarily the scope of this IP. **To me, this is the most expedient choice, because it does not add new complexity to the project** (it mostly means we won't delete the DAG, and that when we write the DAG it should consider that it must be idempotent and can be re-run, which is already a principle of DAG writing, I think).

> We could also consider switching to serving single results from Elasticsearch rather than the API DB to address this. I haven't considered the feasibility of it yet but we'd essentially just need to compare what data we don't index in es and if there would be downsides to adding it (slowing queries, filling the es disk, etc.)

We cannot do that for the reasons I've shared above. ES does not contain all data needed for presentation, and to make it so would require significant new work, far outside the scope of this project or any other project we have on our roadmap, as well as explicitly inverse to previous decisions (see the PR I linked above).